File: src/scripting/ComponentInterop.cpp
#include "ComponentInterop.h"
#include "ecs/Scene.h"
#include "ecs/Components.h"
#include <string>
#include <iostream>
#include <cstring>
#include "DotNetHost.h"
#include "animation/AnimationPlayerComponent.h"
#include "animation/AnimatorRuntime.h"

// --- Global Function Pointers ---
HasComponent_fn HasComponentPtr = &HasComponent;
AddComponent_fn AddComponentPtr = &AddComponent;
RemoveComponent_fn RemoveComponentPtr = &RemoveComponent;
GetLightType_fn GetLightTypePtr = &GetLightType;
SetLightType_fn SetLightTypePtr = &SetLightType;
GetLightColor_fn GetLightColorPtr = &GetLightColor;
SetLightColor_fn SetLightColorPtr = &SetLightColor;
GetLightIntensity_fn GetLightIntensityPtr = &GetLightIntensity;
SetLightIntensity_fn SetLightIntensityPtr = &SetLightIntensity;
GetRigidBodyMass_fn GetRigidBodyMassPtr = &GetRigidBodyMass;
SetRigidBodyMass_fn SetRigidBodyMassPtr = &SetRigidBodyMass;
GetRigidBodyIsKinematic_fn GetRigidBodyIsKinematicPtr = &GetRigidBodyIsKinematic;
SetRigidBodyIsKinematic_fn SetRigidBodyIsKinematicPtr = &SetRigidBodyIsKinematic;
GetRigidBodyLinearVelocity_fn GetRigidBodyLinearVelocityPtr = &GetRigidBodyLinearVelocity;
SetRigidBodyLinearVelocity_fn SetRigidBodyLinearVelocityPtr = &SetRigidBodyLinearVelocity;
GetRigidBodyAngularVelocity_fn GetRigidBodyAngularVelocityPtr = &GetRigidBodyAngularVelocity;
SetRigidBodyAngularVelocity_fn SetRigidBodyAngularVelocityPtr = &SetRigidBodyAngularVelocity;
SetBlendShapeWeight_fn SetBlendShapeWeightPtr = &SetBlendShapeWeight;
GetBlendShapeWeight_fn GetBlendShapeWeightPtr = &GetBlendShapeWeight;
GetBlendShapeCount_fn GetBlendShapeCountPtr = &GetBlendShapeCount;
GetBlendShapeName_fn GetBlendShapeNamePtr = &GetBlendShapeName;

// Helper to get entity data, returns nullptr if not found
static EntityData* GetEntityDataHelper(int entityID) {
    return Scene::Get().GetEntityData(entityID);
}

// --- Component Lifetime ---

bool HasComponent(int entityID, const char* componentName) {
    EntityData* data = GetEntityDataHelper(entityID);
    if (!data) return false;

    if (strcmp(componentName, "LightComponent") == 0) return data->Light != nullptr;
    if (strcmp(componentName, "RigidBodyComponent") == 0) return data->RigidBody != nullptr;
    if (strcmp(componentName, "MeshComponent") == 0) return data->Mesh != nullptr;
    if (strcmp(componentName, "Animator") == 0 || strcmp(componentName, "AnimationPlayerComponent") == 0) return data->AnimationPlayer != nullptr;
    // Add other components here...

    return false;
}

void AddComponent(int entityID, const char* componentName) {
    EntityData* data = GetEntityDataHelper(entityID);
    if (!data) return;

    if (strcmp(componentName, "LightComponent") == 0 && !data->Light) {
        std::cout << "[Interop] AddComponent Light -> Entity " << entityID << "\n";
        data->Light = new LightComponent();
    } else if (strcmp(componentName, "RigidBodyComponent") == 0 && !data->RigidBody) {
        data->RigidBody = new RigidBodyComponent();
        if(data->Collider) // If collider exists, create physics body
        {
            Scene::Get().CreatePhysicsBody(entityID, data->Transform, *data->Collider);
        }
    } else if ((strcmp(componentName, "Animator") == 0 || strcmp(componentName, "AnimationPlayerComponent") == 0) && !data->AnimationPlayer) {
        data->AnimationPlayer = new cm::animation::AnimationPlayerComponent();
    }
    // Add other components here...
}

void RemoveComponent(int entityID, const char* componentName) {
    EntityData* data = GetEntityDataHelper(entityID);
    if (!data) return;

    if (strcmp(componentName, "LightComponent") == 0) {
        delete data->Light;
        data->Light = nullptr;
    } else if (strcmp(componentName, "RigidBodyComponent") == 0) {
        Scene::Get().DestroyPhysicsBody(entityID);
        delete data->RigidBody;
        data->RigidBody = nullptr;
    } else if (strcmp(componentName, "Animator") == 0 || strcmp(componentName, "AnimationPlayerComponent") == 0) {
        delete data->AnimationPlayer;
        data->AnimationPlayer = nullptr;
    }
    // Add other components here...
}

// --- LightComponent ---

int GetLightType(int entityID) {
    EntityData* data = GetEntityDataHelper(entityID);
    return (data && data->Light) ? static_cast<int>(data->Light->Type) : 0;
}

void SetLightType(int entityID, int type) {
    EntityData* data = GetEntityDataHelper(entityID);
    if (data && data->Light) {
        data->Light->Type = static_cast<LightType>(type);
        std::cout << "[Interop] SetLightType entity=" << entityID << " type=" << type << "\n";
    }
}

void GetLightColor(int entityID, float* r, float* g, float* b) {
    EntityData* data = GetEntityDataHelper(entityID);
    if (data && data->Light) {
        *r = data->Light->Color.r;
        *g = data->Light->Color.g;
        *b = data->Light->Color.b;
    }
}

void SetLightColor(int entityID, float r, float g, float b) {
    EntityData* data = GetEntityDataHelper(entityID);
    if (data && data->Light) {
        data->Light->Color = {r, g, b};
        std::cout << "[Interop] SetLightColor entity=" << entityID << " color=(" << r << "," << g << "," << b << ")\n";
    }
}

float GetLightIntensity(int entityID) {
    EntityData* data = GetEntityDataHelper(entityID);
    return (data && data->Light) ? data->Light->Intensity : 0.0f;
}

void SetLightIntensity(int entityID, float intensity) {
    EntityData* data = GetEntityDataHelper(entityID);
    if (data && data->Light) data->Light->Intensity = intensity;
}

// --- RigidBodyComponent ---

float GetRigidBodyMass(int entityID) {
    EntityData* data = GetEntityDataHelper(entityID);
    return (data && data->RigidBody) ? data->RigidBody->Mass : 0.0f;
}

void SetRigidBodyMass(int entityID, float mass) {
    EntityData* data = GetEntityDataHelper(entityID);
    if (data && data->RigidBody) {
        data->RigidBody->Mass = mass;
        // Jolt needs to be updated if mass changes while running
    }
}

bool GetRigidBodyIsKinematic(int entityID) {
    EntityData* data = GetEntityDataHelper(entityID);
    return (data && data->RigidBody) ? data->RigidBody->IsKinematic : false;
}

void SetRigidBodyIsKinematic(int entityID, bool isKinematic) {
    EntityData* data = GetEntityDataHelper(entityID);
    if (data && data->RigidBody) {
        data->RigidBody->IsKinematic = isKinematic;
        // Jolt needs to be updated if this changes while running
    }
}

void GetRigidBodyLinearVelocity(int entityID, float* x, float* y, float* z)
{
    EntityData* data = GetEntityDataHelper(entityID);
    if (data && data->RigidBody)
    {
        JPH::BodyID bodyID = data->RigidBody->BodyID;
        if (!bodyID.IsInvalid())
        {
            auto vel = Physics::GetBodyInterface().GetLinearVelocity(bodyID);
            *x = vel.GetX();
            *y = vel.GetY();
            *z = vel.GetZ();
            return;
        }
    }
    *x = *y = *z = 0;
}

void SetRigidBodyLinearVelocity(int entityID, float x, float y, float z)
{
    SetLinearVelocity(entityID, x, y, z); // From EntityInterop
}

void GetRigidBodyAngularVelocity(int entityID, float* x, float* y, float* z)
{
    EntityData* data = GetEntityDataHelper(entityID);
    if (data && data->RigidBody)
    {
        JPH::BodyID bodyID = data->RigidBody->BodyID;
        if (!bodyID.IsInvalid())
        {
            auto vel = Physics::GetBodyInterface().GetAngularVelocity(bodyID);
            *x = vel.GetX();
            *y = vel.GetY();
            *z = vel.GetZ();
            return;
        }
    }
    *x = *y = *z = 0;
}

void SetRigidBodyAngularVelocity(int entityID, float x, float y, float z)
{
    SetAngularVelocity(entityID, x, y, z); // From EntityInterop
}

// --- BlendShapeComponent ---

void SetBlendShapeWeight(int entityID, const char* shapeName, float weight)
{
    EntityData* data = GetEntityDataHelper(entityID);
    if (!data || !data->BlendShapes || !shapeName) return;
    for (auto& shape : data->BlendShapes->Shapes)
    {
        if (shape.Name == shapeName)
        {
            shape.Weight = weight;
            data->BlendShapes->Dirty = true;
            break;
        }
    }
}

float GetBlendShapeWeight(int entityID, const char* shapeName)
{
    EntityData* data = GetEntityDataHelper(entityID);
    if (!data || !data->BlendShapes || !shapeName) return 0.0f;
    for (const auto& shape : data->BlendShapes->Shapes)
    {
        if (shape.Name == shapeName)
        {
            return shape.Weight;
        }
    }
    return 0.0f;
}

int GetBlendShapeCount(int entityID)
{
    EntityData* data = GetEntityDataHelper(entityID);
    return (data && data->BlendShapes) ? data->BlendShapes->Shapes.size() : 0;
}

const char* GetBlendShapeName(int entityID, int index)
{
    EntityData* data = GetEntityDataHelper(entityID);
    if (!data || !data->BlendShapes || index < 0 || index >= data->BlendShapes->Shapes.size())
    {
        return nullptr;
    }
    return data->BlendShapes->Shapes[index].Name.c_str();
}

// --- Animator parameter setters ---
static cm::animation::AnimationPlayerComponent* GetAnimatorFor(int entityID)
{
    EntityData* data = GetEntityDataHelper(entityID);
    return (data ? data->AnimationPlayer : nullptr);
}

void Animator_SetBool(int entityID, const char* name, bool value)
{
    if (!name) return;
    if (auto* ap = GetAnimatorFor(entityID)) {
        ap->AnimatorInstance.Blackboard().Bools[name] = value;
    }
}

void Animator_SetInt(int entityID, const char* name, int value)
{
    if (!name) return;
    if (auto* ap = GetAnimatorFor(entityID)) {
        ap->AnimatorInstance.Blackboard().Ints[name] = value;
    }
}

void Animator_SetFloat(int entityID, const char* name, float value)
{
    if (!name) return;
    if (auto* ap = GetAnimatorFor(entityID)) {
        ap->AnimatorInstance.Blackboard().Floats[name] = value;
    }
}

void Animator_SetTrigger(int entityID, const char* name)
{
    if (!name) return;
    if (auto* ap = GetAnimatorFor(entityID)) {
        ap->AnimatorInstance.Blackboard().Triggers[name] = true;
    }
}

void Animator_ResetTrigger(int entityID, const char* name)
{
    if (!name) return;
    if (auto* ap = GetAnimatorFor(entityID)) {
        ap->AnimatorInstance.Blackboard().Triggers[name] = false;
    }
}

// --- AnimationPlayer (single-clip mode) controls ---
void AnimationPlayer_Play(int entityID)
{
    if (auto* ap = GetAnimatorFor(entityID)) {
        ap->AnimatorMode = cm::animation::AnimationPlayerComponent::Mode::AnimationPlayerAnimated;
        ap->IsPlaying = true;
        if (!ap->ActiveStates.empty()) ap->ActiveStates.front().Time = 0.0f;
    }
}

void AnimationPlayer_Stop(int entityID)
{
    if (auto* ap = GetAnimatorFor(entityID)) {
        ap->IsPlaying = false;
    }
}

bool AnimationPlayer_IsPlaying(int entityID)
{
    if (auto* ap = GetAnimatorFor(entityID)) return ap->IsPlaying;
    return false;
}

void AnimationPlayer_SetLoop(int entityID, bool loop)
{
    if (auto* ap = GetAnimatorFor(entityID)) {
        if (ap->ActiveStates.empty()) ap->ActiveStates.push_back({});
        ap->ActiveStates.front().Loop = loop;
    }
}

void AnimationPlayer_SetSpeed(int entityID, float speed)
{
    if (auto* ap = GetAnimatorFor(entityID)) ap->PlaybackSpeed = speed;
}

const char* AnimationPlayer_GetCurrentClipName(int entityID)
{
    if (auto* ap = GetAnimatorFor(entityID)) return ap->Debug_CurrentAnimationName.c_str();
    return "";
}

const char* Animator_GetCurrentStateName(int entityID)
{
    if (auto* ap = GetAnimatorFor(entityID)) return ap->Debug_CurrentControllerStateName.c_str();
    return "";
}

bool Animator_IsPlaying(int entityID)
{
    if (auto* ap = GetAnimatorFor(entityID)) {
        if (ap->AnimatorMode == cm::animation::AnimationPlayerComponent::Mode::ControllerAnimated)
            return true; // controller is always advancing
        return ap->IsPlaying;
    }
    return false;
}

File: scripts/ClaymoreEngine/ComponentInterop.cs
using System;
using System.Numerics;
using System.Runtime.InteropServices;

namespace ClaymoreEngine
{
    // These delegates must match the C++ function pointer types in DotNetHost.h
    
    // Component Lifetime
    [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
    internal delegate bool HasComponentFn(int entityId, [MarshalAs(UnmanagedType.LPStr)] string componentName);

    [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
    internal delegate void AddComponentFn(int entityId, [MarshalAs(UnmanagedType.LPStr)] string componentName);

    [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
    internal delegate void RemoveComponentFn(int entityId, [MarshalAs(UnmanagedType.LPStr)] string componentName);

    // LightComponent
    [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
    internal delegate int GetLightTypeFn(int entityId);

    [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
    internal delegate void SetLightTypeFn(int entityId, int type);

    [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
    internal delegate void GetLightColorFn(int entityId, out float r, out float g, out float b);

    [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
    internal delegate void SetLightColorFn(int entityId, float r, float g, float b);
    
    [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
    internal delegate float GetLightIntensityFn(int entityId);

    [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
    internal delegate void SetLightIntensityFn(int entityId, float intensity);

    // RigidBodyComponent
    [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
    internal delegate float GetRigidBodyMassFn(int entityId);
    
    [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
    internal delegate void SetRigidBodyMassFn(int entityId, float mass);

    [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
    internal delegate bool GetRigidBodyIsKinematicFn(int entityId);

    [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
    internal delegate void SetRigidBodyIsKinematicFn(int entityId, bool isKinematic);

    [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
    internal delegate void GetRigidBodyLinearVelocityFn(int entityId, out float x, out float y, out float z);

    [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
    internal delegate void SetRigidBodyLinearVelocityFn(int entityId, float x, float y, float z);
    
    [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
    internal delegate void GetRigidBodyAngularVelocityFn(int entityId, out float x, out float y, out float z);

    [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
    internal delegate void SetRigidBodyAngularVelocityFn(int entityId, float x, float y, float z);

    // BlendShapeComponent
    [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
    internal delegate void SetBlendShapeWeightFn(int entityId, [MarshalAs(UnmanagedType.LPStr)] string shapeName, float weight);
    [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
    internal delegate float GetBlendShapeWeightFn(int entityId, [MarshalAs(UnmanagedType.LPStr)] string shapeName);
    [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
    internal delegate int GetBlendShapeCountFn(int entityId);
    [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
    internal delegate IntPtr GetBlendShapeNameFn(int entityId, int index);

    // Animator parameter setters
    [UnmanagedFunctionPointer(CallingConvention.Cdecl)] internal delegate void Animator_SetBoolFn(int entityId, [MarshalAs(UnmanagedType.LPStr)] string name, bool value);
    [UnmanagedFunctionPointer(CallingConvention.Cdecl)] internal delegate void Animator_SetIntFn(int entityId, [MarshalAs(UnmanagedType.LPStr)] string name, int value);
    [UnmanagedFunctionPointer(CallingConvention.Cdecl)] internal delegate void Animator_SetFloatFn(int entityId, [MarshalAs(UnmanagedType.LPStr)] string name, float value);
    [UnmanagedFunctionPointer(CallingConvention.Cdecl)] internal delegate void Animator_SetTriggerFn(int entityId, [MarshalAs(UnmanagedType.LPStr)] string name);
    [UnmanagedFunctionPointer(CallingConvention.Cdecl)] internal delegate void Animator_ResetTriggerFn(int entityId, [MarshalAs(UnmanagedType.LPStr)] string name);
    // Animator getters
    [UnmanagedFunctionPointer(CallingConvention.Cdecl)] internal delegate bool  Animator_GetBoolFn(int entityId, [MarshalAs(UnmanagedType.LPStr)] string name);
    [UnmanagedFunctionPointer(CallingConvention.Cdecl)] internal delegate int   Animator_GetIntFn(int entityId, [MarshalAs(UnmanagedType.LPStr)] string name);
    [UnmanagedFunctionPointer(CallingConvention.Cdecl)] internal delegate float Animator_GetFloatFn(int entityId, [MarshalAs(UnmanagedType.LPStr)] string name);
    [UnmanagedFunctionPointer(CallingConvention.Cdecl)] internal delegate bool  Animator_GetTriggerFn(int entityId, [MarshalAs(UnmanagedType.LPStr)] string name);


    internal static unsafe class ComponentInterop
    {
        // --- Function pointers ---
        public static HasComponentFn HasComponent;
        public static AddComponentFn AddComponent;
        public static RemoveComponentFn RemoveComponent;

        // Light
        public static GetLightTypeFn GetLightType;
        public static SetLightTypeFn SetLightType;
        public static GetLightColorFn GetLightColor;
        public static SetLightColorFn SetLightColor;
        public static GetLightIntensityFn GetLightIntensity;
        public static SetLightIntensityFn SetLightIntensity;

        // RigidBody
        public static GetRigidBodyMassFn GetRigidBodyMass;
        public static SetRigidBodyMassFn SetRigidBodyMass;
        public static GetRigidBodyIsKinematicFn GetRigidBodyIsKinematic;
        public static SetRigidBodyIsKinematicFn SetRigidBodyIsKinematic;
        public static GetRigidBodyLinearVelocityFn GetRigidBodyLinearVelocity;
        public static SetRigidBodyLinearVelocityFn SetRigidBodyLinearVelocity;
        public static GetRigidBodyAngularVelocityFn GetRigidBodyAngularVelocity;
        public static SetRigidBodyAngularVelocityFn SetRigidBodyAngularVelocity;

        // BlendShapes
        public static SetBlendShapeWeightFn SetBlendShapeWeight;
        public static GetBlendShapeWeightFn GetBlendShapeWeight;
        public static GetBlendShapeCountFn GetBlendShapeCount;
        public static GetBlendShapeNameFn GetBlendShapeNameInternal;

        // Animator
        public static Animator_SetBoolFn Animator_SetBool;
        public static Animator_SetIntFn Animator_SetInt;
        public static Animator_SetFloatFn Animator_SetFloat;
        public static Animator_SetTriggerFn Animator_SetTrigger;
        public static Animator_ResetTriggerFn Animator_ResetTrigger;
        public static Animator_GetBoolFn Animator_GetBool;
        public static Animator_GetIntFn Animator_GetInt;
        public static Animator_GetFloatFn Animator_GetFloat;
        public static Animator_GetTriggerFn Animator_GetTrigger;

        public static string GetBlendShapeName(int entityId, int index)
        {
            IntPtr ptr = GetBlendShapeNameInternal(entityId, index);
            return Marshal.PtrToStringAnsi(ptr);
        }

        public static void Initialize(void** ptrs, int count)
        {
            if (count < 30) // Update count for added animator functions
            {
                Console.WriteLine($"[ComponentInterop] Expected at least 30 function pointers, but got {count}.");
                return;
            }

            int i = 0;
            HasComponent = Marshal.GetDelegateForFunctionPointer<HasComponentFn>((IntPtr)ptrs[i++]);
            AddComponent = Marshal.GetDelegateForFunctionPointer<AddComponentFn>((IntPtr)ptrs[i++]);
            RemoveComponent = Marshal.GetDelegateForFunctionPointer<RemoveComponentFn>((IntPtr)ptrs[i++]);
            
            GetLightType = Marshal.GetDelegateForFunctionPointer<GetLightTypeFn>((IntPtr)ptrs[i++]);
            SetLightType = Marshal.GetDelegateForFunctionPointer<SetLightTypeFn>((IntPtr)ptrs[i++]);
            GetLightColor = Marshal.GetDelegateForFunctionPointer<GetLightColorFn>((IntPtr)ptrs[i++]);
            SetLightColor = Marshal.GetDelegateForFunctionPointer<SetLightColorFn>((IntPtr)ptrs[i++]);
            GetLightIntensity = Marshal.GetDelegateForFunctionPointer<GetLightIntensityFn>((IntPtr)ptrs[i++]);
            SetLightIntensity = Marshal.GetDelegateForFunctionPointer<SetLightIntensityFn>((IntPtr)ptrs[i++]);

            GetRigidBodyMass = Marshal.GetDelegateForFunctionPointer<GetRigidBodyMassFn>((IntPtr)ptrs[i++]);
            SetRigidBodyMass = Marshal.GetDelegateForFunctionPointer<SetRigidBodyMassFn>((IntPtr)ptrs[i++]);
            GetRigidBodyIsKinematic = Marshal.GetDelegateForFunctionPointer<GetRigidBodyIsKinematicFn>((IntPtr)ptrs[i++]);
            SetRigidBodyIsKinematic = Marshal.GetDelegateForFunctionPointer<SetRigidBodyIsKinematicFn>((IntPtr)ptrs[i++]);
            GetRigidBodyLinearVelocity = Marshal.GetDelegateForFunctionPointer<GetRigidBodyLinearVelocityFn>((IntPtr)ptrs[i++]);
            SetRigidBodyLinearVelocity = Marshal.GetDelegateForFunctionPointer<SetRigidBodyLinearVelocityFn>((IntPtr)ptrs[i++]);
            GetRigidBodyAngularVelocity = Marshal.GetDelegateForFunctionPointer<GetRigidBodyAngularVelocityFn>((IntPtr)ptrs[i++]);
            SetRigidBodyAngularVelocity = Marshal.GetDelegateForFunctionPointer<SetRigidBodyAngularVelocityFn>((IntPtr)ptrs[i++]);

            SetBlendShapeWeight = Marshal.GetDelegateForFunctionPointer<SetBlendShapeWeightFn>((IntPtr)ptrs[i++]);
            GetBlendShapeWeight = Marshal.GetDelegateForFunctionPointer<GetBlendShapeWeightFn>((IntPtr)ptrs[i++]);
            GetBlendShapeCount = Marshal.GetDelegateForFunctionPointer<GetBlendShapeCountFn>((IntPtr)ptrs[i++]);
            GetBlendShapeNameInternal = Marshal.GetDelegateForFunctionPointer<GetBlendShapeNameFn>((IntPtr)ptrs[i++]);

            Animator_SetBool = Marshal.GetDelegateForFunctionPointer<Animator_SetBoolFn>((IntPtr)ptrs[i++]);
            Animator_SetInt = Marshal.GetDelegateForFunctionPointer<Animator_SetIntFn>((IntPtr)ptrs[i++]);
            Animator_SetFloat = Marshal.GetDelegateForFunctionPointer<Animator_SetFloatFn>((IntPtr)ptrs[i++]);
            Animator_SetTrigger = Marshal.GetDelegateForFunctionPointer<Animator_SetTriggerFn>((IntPtr)ptrs[i++]);
            Animator_ResetTrigger = Marshal.GetDelegateForFunctionPointer<Animator_ResetTriggerFn>((IntPtr)ptrs[i++]);
            Animator_GetBool = Marshal.GetDelegateForFunctionPointer<Animator_GetBoolFn>((IntPtr)ptrs[i++]);
            Animator_GetInt = Marshal.GetDelegateForFunctionPointer<Animator_GetIntFn>((IntPtr)ptrs[i++]);
            Animator_GetFloat = Marshal.GetDelegateForFunctionPointer<Animator_GetFloatFn>((IntPtr)ptrs[i++]);
            Animator_GetTrigger = Marshal.GetDelegateForFunctionPointer<Animator_GetTriggerFn>((IntPtr)ptrs[i++]);
        }
    }
}

